pyteal/ast/abi/type.py

```diff
@@ -76,9 +77,27 @@ class BaseType(ABC):
 
     def __init__(self, spec: TypeSpec) -> None:
         """Create a new BaseType."""
+        from pyteal.ast.subroutine import SubroutineEval
+
         super().__init__()
         self._type_spec: Final[TypeSpec] = spec
-        self._stored_value: AbstractVar = ScratchVar(spec.storage_type())
+        self._stored_value: AbstractVar
+
+        if SubroutineEval.Context.proto:
+            proto = SubroutineEval.Context.proto
+
+            assert proto.mem_layout
```

Comment:
Another good point, I have to admit that letting `proto.mem_layout` to be okay with `None` is not so well considered...

> Perhaps the proto member of Context object is constructed in a way that guarantees that it's well-formed?

yes, that is correct, for the proto generated in subroutine is guaranteed to be well-formed. But still, I think it might be better to let `proto.mem_layout` to be not `None`.

Reply:
Pushed the change in [06774f6](https://github.com/algorand/pyteal/pull/606/commits/06774f675f208f5085760014ed20929a9bd7129d)